server/src/main/java/org/candlepin/model/EntitlementCurator.java

```diff
@@ -302,39 +271,31 @@ private Criteria createModifiesDateFilteringCriteria(Set<Consumer> consumers, Da
-        Set<Entitlement> modifying = new HashSet<Entitlement>();
-        // Get the map of product Ids to the set of
-        // overlapping entitlements that provide them
-        ProductEntitlements pidEnts = getOverlappingForModifying(entitlements);
-        if (pidEnts.isEmpty()) {
-            // Empty collections break hibernate queries
-            return modifying;
-        }
-
-        List<Product> overlappingProducts = productCurator.listAllByIds(pidEnts.getAllProductIds());
-
-        Set<String> entitlementProducts = new HashSet<String>();
-        for (Entitlement entitlement : entitlements) {
-            entitlementProducts.add(entitlement.getPool().getProductId());
-            for (Product product : entitlement.getPool().getProvidedProducts()) {
-                entitlementProducts.add(product.getId());
-            }
-        }
-
-        for (Product overlappingProduct : overlappingProducts) {
-            boolean modifies = false;
-            Iterator<String> ppit = entitlementProducts.iterator();
-            // No need to continue checking once we have found a modified product
-            while (!modifies && ppit.hasNext()) {
-                modifies = overlappingProduct.modifies(ppit.next());
-            }
-            if (modifies) {
-                // Return all entitlements for the modified product
-                modifying.addAll(pidEnts.getEntitlementsByProductId(overlappingProduct.getId()));
-            }
-        }
-
-        return modifying;
+        List<Entitlement> ents  = getEntityManager().createQuery(
+            "SELECT DISTINCT eOut " +
+            "    FROM Entitlement eOut  " +
+            "     JOIN eOut.pool outPool " +
+            "     JOIN outPool.providedProducts outProvided " +
+            "     JOIN outProvided.productContent outProvContent  " +
+            "     JOIN outProvContent.content outContent" +
+            "     JOIN outContent.modifiedProductIds outModProdId " +
+            "    WHERE " +
+            "     outPool.endDate >= current_date AND" +
+            "     eOut NOT IN (:ein) AND" +
+            "     EXISTS (SELECT eIn " +
+            "             FROM Entitlement eIn " +
+            "              JOIN eIn.consumer inConsumer " +
+            "              JOIN eIn.pool inPool " +
+            "              JOIN inPool.product inMktProd " +
+            "              LEFT JOIN inPool.providedProducts inProvidedProd" +
+            "              WHERE eIn in (:ein) AND inConsumer = eOut.consumer AND " +
+            "                inPool.endDate >= outPool.startDate AND " +
+            "                inPool.startDate <= outPool.endDate AND" +
+            "                ( inProvidedProd.id = outModProdId OR inMktProd.id = outModProdId ) " +
+            "             )" +
+            "                       "
+            , Entitlement.class).setParameter("ein", entitlements).getResultList();
+        return new HashSet<Entitlement>(ents);
```

Comment:
The only place we use this method is in the block above in CPM. We should probably just change the return type to Collection<Entitlement> rather than copying everything to a new set.

Also, with the changes forth-coming in my branch, it might be better to change this to return just the entitlement IDs rather than the objects (or, at least, have a version of this method which does so).